[Cocoa] Add telemetry for load failures caused by DNS resolution errors

whsieh opened the pull request:
#### d24e5f4ed53b371a945f324805758edccb55b73f
<pre>
[Cocoa] Add telemetry for load failures caused by DNS resolution errors
<a href="https://bugs.webkit.org/show_bug.cgi?id=251742">https://bugs.webkit.org/show_bug.cgi?id=251742</a>
rdar://104767598

Reviewed by Tim Horton.

Surface a dictionary containing more information about DNS resolution failures in the network
process, through the UI-process side navigation delegate. This dictionary contains basic information
derived from the DNS resolution failure report (`nw_resolution_report_t`) as well as path
information (`nw_path_t`) that&apos;s bubbled up to WebKit, through CFNetwork.

Clients can use this information for debugging purposes, as well as providing additional telemetry.
See below for more details.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(description):
(extractResolutionReport):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::isSafeToEncodeUserInfo):

Currently, when serializing `NSError`s, we only attempt to serialize values in the user info
dictionary that are one of the following types that are safe to encode: `NSURL`, `NSString`, or
`NSNumber`. In order to propagate the new user info entry above to the UI process, we augment this
rule so that we also serialize Objective-C dictionaries and arrays that, in turn, only contain data
types that are safe to encode.

(IPC::encodeNSError):

Canonical link: <a href="https://commits.webkit.org/259908@main">https://commits.webkit.org/259908@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bf6e432caefdd1409c08c4066b79f4551419f00

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106373 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15421 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39209 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115566 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6643 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98591 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115243 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112138 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95819 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40397 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27462 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8651 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28813 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9173 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5387 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14802 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48360 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6847 "'git push ...'") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10730 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -805,10 +870,20 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
         NSDictionary *oldUserInfo = [error userInfo];
         NSMutableDictionary *newUserInfo = oldUserInfo ? [NSMutableDictionary dictionaryWithDictionary:oldUserInfo] : [NSMutableDictionary dictionary];
         newUserInfo[@"networkTaskDescription"] = [task description];
+        if (auto* networkDataTask = [self existingTask:task]) {
```
Diff looks wrong, no? (see two lines down)

whsieh replied:
Fixed — it looks like I applied the diff incorrectly when I rebased, oops.

Thanks for the review!

whsieh committed:
[Cocoa] Add telemetry for load failures caused by DNS resolution errors
https://bugs.webkit.org/show_bug.cgi?id=251742
rdar://104767598

Reviewed by Tim Horton.

Surface a dictionary containing more information about DNS resolution failures in the network
process, through the UI-process side navigation delegate. This dictionary contains basic information
derived from the DNS resolution failure report (`nw_resolution_report_t`) as well as path
information (`nw_path_t`) that's bubbled up to WebKit, through CFNetwork.

Clients can use this information for debugging purposes, as well as providing additional telemetry.
See below for more details.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(description):
(extractResolutionReport):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::isSafeToEncodeUserInfo):

Currently, when serializing `NSError`s, we only attempt to serialize values in the user info
dictionary that are one of the following types that are safe to encode: `NSURL`, `NSString`, or
`NSNumber`. In order to propagate the new user info entry above to the UI process, we augment this
rule so that we also serialize Objective-C dictionaries and arrays that, in turn, only contain data
types that are safe to encode.

(IPC::encodeNSError):

Canonical link: https://commits.webkit.org/259908@main
